File: api/client/src/main/java/su/plo/voice/api/client/audio/device/AudioDevice.java
package su.plo.voice.api.client.audio.device;


import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import su.plo.voice.api.client.audio.filter.AudioFilter;
import su.plo.voice.api.util.Params;

import javax.sound.sampled.AudioFormat;
import java.util.Collection;
import java.util.Optional;
import java.util.function.Predicate;

// todo: doc
public interface AudioDevice {

    /**
     * Opens the device
     *
     * @param format the audio format
     * @param params device params
     *
     * @throws DeviceException if the device cannot be open
     */
    void open(@NotNull AudioFormat format, @NotNull Params params) throws DeviceException;

    /**
     * Reloads the device
     */
    void reload() throws DeviceException;

    /**
     * Closes the device
     */
    void close();

    /**
     * @return true if the device is open
     */
    boolean isOpen();

    /**
     * Adds the filter to device with priority
     */
    void addFilter(AudioFilter filter, AudioFilter.Priority priority);

    /**
     * Adds the filter to device with normal priority
     */
    void addFilter(AudioFilter filter);

    /**
     * Removes the filter from device
     */
    void removeFilter(AudioFilter filter);

    /**
     * Gets the device filters
     *
     * @return the device filters
     */
    Collection<AudioFilter> getFilters();

    /**
     * Process all filters
     *
     * @return the processed samples
     */
    short[] processFilters(short[] samples, @Nullable Predicate<AudioFilter> excludeFilter);

    /**
     * Process all filters
     *
     * @return the processed samples
     */
    default short[] processFilters(short[] samples) {
        return processFilters(samples, null);
    }

    /**
     * Gets the device name
     *
     * @return the device name
     */
    String getName();

    /**
     * Gets the device audio format
     *
     * @return the audio format with which audio device was open
     */
    Optional<AudioFormat> getFormat();

    /**
     * Gets the device params
     *
     * @return the device params
     */
    Optional<Params> getParams();

    /**
     * Gets the device buffer size calculated from format
     *
     * @return the buffer size
     */
    int getBufferSize();

    /**
     * Gets the device type
     *
     * @return the device type
     */
    DeviceType getType();
}

File: api/client/src/main/kotlin/su/plo/voice/api/client/audio/device/AudioDevice.kt
package su.plo.voice.api.client.audio.device

import su.plo.voice.api.client.audio.filter.AudioFilter
import java.util.*
import java.util.function.Predicate
import javax.sound.sampled.AudioFormat

/**
 * Represents an audio device used for audio input or output.
 */
interface AudioDevice {

    /**
     * Gets the name of the audio device.
     *
     * @return The name of the device.
     */
    val name: String

    /**
     * Gets the audio format associated with the device.
     *
     * @return The audio format with which the audio device was opened.
     */
    val format: AudioFormat

    /**
     * Gets the 20ms frame size calculated based on the audio format.
     *
     * @return The frame size for the device.
     */
    val frameSize: Int

    /**
     * Gets the type of the audio device.
     *
     * @return The type of the device.
     */
    val type: DeviceType

    /**
     * Reloads (closes and opens) the audio device.
     *
     * @throws DeviceException if there is an issue reloading the device.
     */
    @Throws(DeviceException::class)
    fun reload()

    /**
     * Closes the audio device.
     */
    fun close()

    /**
     * Checks if the audio device is open.
     *
     * @return `true` if the device is open, `false` otherwise.
     */
    fun isOpen(): Boolean

    /**
     * Adds an audio filter to the device with the specified priority.
     *
     * @param filter   The audio filter to add.
     * @param priority The priority of the filter.
     */
    fun addFilter(filter: AudioFilter, priority: AudioFilter.Priority)

    /**
     * Adds an audio filter to the device with normal priority.
     *
     * @param filter The audio filter to add.
     */
    fun addFilter(filter: AudioFilter) {
        addFilter(filter, AudioFilter.Priority.NORMAL)
    }

    /**
     * Removes an audio filter from the device.
     *
     * @param filter The audio filter to remove.
     */
    fun removeFilter(filter: AudioFilter)

    /**
     * Gets the collection of audio filters applied to the device.
     *
     * @return The collection of audio filters.
     */
    fun getFilters(): Collection<AudioFilter>

    /**
     * Processes the provided audio samples through all applicable filters.
     *
     * @param samples       The input audio samples to process.
     * @param excludeFilter A filter predicate to exclude specific filters from processing.
     * @return The processed audio samples.
     */
    fun processFilters(samples: ShortArray, excludeFilter: Predicate<AudioFilter>?): ShortArray

    /**
     * Processes the provided audio samples through all applicable filters.
     *
     * @param samples The input audio samples to process.
     * @return The processed audio samples.
     */
    fun processFilters(samples: ShortArray): ShortArray {
        return processFilters(samples, null)
    }
}
